### ISSUE TITLE
Use light-rag without cuda

### ISSUE THREAD

**Issue by Fianax** [HKUDS/LightRAG#454]
Can light-rag be used without cuda?
Using only cpu?

or if your working environment has no gpu you can't use it? even if it takes longer, I wouldn't mind.

**Comment by alllexx88** [comment]
You can use ollama for LLM and embeddings, and ollama works locally on CPU only as well, though surely slower. Based on `lightrag_ollama_demo.py`, this should work:

```python
import asyncio
import os
import inspect
import logging
from lightrag import LightRAG, QueryParam
from lightrag.llm import ollama_model_complete, ollama_embedding
from lightrag.utils import EmbeddingFunc

WORKING_DIR = "./dickens"

logging.basicConfig(format="%(levelname)s:%(message)s", level=logging.INFO)

if not os.path.exists(WORKING_DIR):
    os.mkdir(WORKING_DIR)

rag = LightRAG(
    working_dir=WORKING_DIR,
    llm_model_func=ollama_model_complete,
    llm_model_name="llama3.1:8b",
    llm_model_max_async=4,
    llm_model_max_token_size=32768,
    llm_model_kwargs={"host": "http://localhost:11434", "options": {"num_ctx": 32768}},
    embedding_func=EmbeddingFunc(
        embedding_dim=768,
        max_token_size=8192,
        func=lambda texts: ollama_embedding(
            texts, embed_model="nomic-embed-text", host="http://localhost:11434"
        ),
    ),
)

with open("./book.txt", "r", encoding="utf-8") as f:
    rag.insert(f.read())

# Perform hybrid search
print(
    rag.query("What are Scrooge's greatest fears?", param=QueryParam(mode="hybrid"))
)
```

Make sure you have ollama installed and started, as well as `llama3.1:8b` and `nomic-embed-text` models pulled.

State (kv, vector and graph storages etc.) will be stored in `WORKING_DIR = "./dickens"`, you can then run this code again without the call to `rag.insert()` with new queries.

`book.txt` per the readme can be downloaded with:
```sh
curl https://raw.githubusercontent.com/gusye1234/nano-graphrag/main/tests/mock_data.txt > ./book.txt
```

**Comment by Fianax** [comment]
Thank you very much for the example and the explanation.

I just wanted to try it locally, even if it is slow, to see the results.

Would the same work with the huggenface example? Because I can infer their models locally with cpu (although they are slow) and I wanted to test it.

**Comment by alllexx88** [comment]
You're welcome :smile: 
Huggingface models should run on CPU just fine, but I think ollama is still faster. But you can test yourself :smiley:

**Comment by Fianax** [comment]
but that is not the problem.
what happens is that if I run the example you have of huggingface it fails me because it tries to use gpu when I don't have gpu in my computer.

that's why i was asking how to enable the execution of the example you have with cpu and not with gpu.

**Comment by alllexx88** [comment]
This seems to be the only place in lightrag code to use cuda explicitly:
https://github.com/HKUDS/LightRAG/blob/3e33cfc20effaada8ebcc0b26a9d261165b09649/lightrag/llm.py#L291

You could remove `.to("cuda")` but I'm not sure if it'll work. It's easier to just use ollama instead.

**Comment by Fianax** [comment]
and if I use ollama, even if I don't have ‘cuda’, it won't fail me because of ‘cuda’, right?

**Comment by alllexx88** [comment]
Yes it won't, that `cuda` snippet is only relevant for hf models, and ollama works without cuda just fine, the client API has no control over which runner ollama actually uses for the inference.
